Complete implementation of PNG IHDRC chunk decoding

piponazo opened the pull request:
I started digging into the PNG specification and I will create few PRs dedicated to improve the decoding of PNG chunks. 

In this first PR I just focus on the [IHDR chunk](http://www.libpng.org/pub/png/spec/1.2/png-1.2-pdg.html#C.IHDR). We were only reading the width and height values from there and ignoring the rest of values. Now we read all the fields and we do the pertinent checks on their values. This should make the code more robust against malicious handmade PNG files. 

I also took the opportunity to continue updating the doxygen style and clang-format the files touched. 

Recommendation: review commit by commit.

piponazo committed:
Update doxygen style

piponazo committed:
Move globals into anonymous namespace

piponazo committed:
clang-format in PNG files

piponazo committed:
No need to check for LIBZ (it is required in CMake if PNG is enabled)

piponazo committed:
Replace EXV_HAVE_LIBZ with EXV_HAVE_PNG

piponazo committed:
Update doxygen style in pngchunk_int.hpp

piponazo committed:
clang-format in pngchunk_int.hpp

piponazo committed:
clang-format in pngchunk_int.cpp

piponazo committed:
Add unit tests for PNG IHDR chunk decoding

piponazo committed:
Complete PNG IHDR-chunk decoding implementation

piponazo committed:
Update test case. Now a PNG TOC throws because of corrupted metadata

D4N commented:
@piponazo Could you rebase this on master so that we get a coverage report?

piponazo commented:
> @piponazo Could you rebase this on master so that we get a coverage report?

done

clanmills reviewed (commented):
This looks like a good overhaul of the PNG code.  And that's a good thing.  Thanks for digging around in that.

I am concerned about the symbol EXV_HAVE_PNG as that implies we are using libpng.  Perhaps EXV_ENABLE_PNG is a better name which (based on EXIV2_ENABLE_PNG in CMakeLists.txt).

For that matter, perhaps it's worthwhile replacing EXV_this_that with EXIV2_this_that.

piponazo commented:
> I am concerned about the symbol EXV_HAVE_PNG as that implies we are using libpng. Perhaps EXV_ENABLE_PNG is a better name which (based on EXIV2_ENABLE_PNG in CMakeLists.txt). For that matter, perhaps it's worthwhile replacing EXV_this_that with EXIV2_this_that.

I totally agree with you. I'll make both changes 😉

D4N commented on src/pngchunk_int.hpp:
```diff
@@ -20,186 +20,136 @@
+            static std::string makeAsciiTxtChunk(const std::string& keyword, const std::string& text, bool compress);
+
+            /// @brief Return a compressed or uncompressed (iTXt) PNG international text chunk (length + chunk type +
+            /// chunk data + CRC) as a string.
```
```suggestion
            ///     chunk data + CRC) as a string.
```

D4N reviewed (changes requested):
I've added some nagging to the documentation, but unfortunately realized too late that you only reformated most of the doc comments, so feel free to ignore the "please clarify this" comments, unless you're highly motivated to address them :wink:.

I'd suggest to throw `kerFailedToReadMetadata` from the function that you completed, for consistency with the other png reading functions (that would also make the changes in the system test obsolete).

piponazo commented on include/exiv2/pngimage.hpp:
```diff
@@ -40,48 +40,32 @@
+        /// If a new image is to be created, any existing data is overwritten. Since the constructor can not return a
+        /// result, callers should check the good() method after object construction to determine success or failure.
+        /// @param io An auto-pointer that owns a BasicIo instance used for reading and writing image metadata.
+        /// \b Important: The constructor takes ownership of the passed in BasicIo instance through the auto-pointer.
```
I agree, I saw some of these comments and I actually simplified them (not sure if it was done in this PR or in other).

piponazo commented on include/exiv2/pngimage.hpp:
```diff
@@ -40,48 +40,32 @@
         void writeMetadata() override;
-
-        /*!
-          @brief Print out the structure of image file.
```
It is exactly the same comment that comes from the Base class (**Image**). In Doxygen, there is documentation inheritance :smile_cat:

piponazo commented on src/pngchunk_int.hpp:
```diff
@@ -20,186 +20,136 @@
+            /// Supported Chunk types are tTXt, zTXt, and iTXt.
+            static DataBuf parseTXTChunk(const DataBuf& data, int keysize, TxtChunkType type);
+
+            /// @brief Parse PNG chunk contents to extract metadata container and assign it to image.
```
I'll try to do it when I write unit tests for it in the future :wink:

D4N committed:
Minor improvements in documentation

Apply suggestions from code review

Co-Authored-By: piponazo <piponazo@gmail.com>

piponazo committed:
Change EXV_HAVE_PNG into EXIV2_ENABLE_PNG

piponazo commented:
> I'd suggest to throw kerFailedToReadMetadata from the function that you completed, for consistency with the other png reading functions (that would also make the changes in the system test obsolete).

I am happy to apply this change, although the truth is that there is an existing mix in that file of exceptions: `kerFailedToReadImageData` and `kerCorruptedMetadata`. Do you know the difference between those 2 values ? When we should use one or the other ? In case there is not so much difference I would vote for removing one of them (In another PR).

piponazo commented:
> For that matter, perhaps it's worthwhile replacing EXV_this_that with EXIV2_this_that.

@clanmills I just replaced the definition I changed in this file. Since to do that change for all the definitions is a quite big task, I would prefer to do it in another PR.

clanmills commented:
Changing EXV_HAVE_PNG to EXIV2_ENABLE_PNG is the change I requested.

Changing all those EXV_ defines is something for another day.  We could ask Nehal to do that when everybody's on vacation (July?) and the change slides into the code without friction.

D4N commented:
> > I'd suggest to throw kerFailedToReadMetadata from the function that you completed, for consistency with the other png reading functions (that would also make the changes in the system test obsolete).
> 
> I am happy to apply this change, although the truth is that there is an existing mix in that file of exceptions: `kerFailedToReadImageData` and `kerCorruptedMetadata`. Do you know the difference between those 2 values ? When we should use one or the other ? In case there is not so much difference I would vote for removing one of them (In another PR).

I think `kerFailedToReadImageData` existed a long time before `kerCorruptedMetadata`, which I might have added when I just started contributing ("back then" we just had error codes and no enum for the numbers, so new errors got added left right and center, see #323)
